v-next/core/src/internal/providers/errors.ts

```diff
@@ -0,0 +1,57 @@
+import type { ProviderRpcError } from "../../types/providers.js";
+
+import { CustomError } from "@ignored/hardhat-vnext-utils/error";
+import { isObject } from "@ignored/hardhat-vnext-utils/lang";
+
+const IS_PROVIDER_ERROR_PROPERTY_NAME = "_isProviderError";
+
+/**
+ * The error codes that a provider can return.
+ * See https://eips.ethereum.org/EIPS/eip-1474#error-codes
+ */
+export enum ProviderErrorCode {
+  LIMIT_EXCEEDED = -32005,
+  INVALID_PARAMS = -32602,
+}
+
+type ProviderErrorMessages = {
+  [key in ProviderErrorCode]: string;
+};
+
+/**
+ * The error messages associated with each error code.
+ */
+const ProviderErrorMessage: ProviderErrorMessages = {
+  [ProviderErrorCode.LIMIT_EXCEEDED]: "Request exceeds defined limit",
+  [ProviderErrorCode.INVALID_PARAMS]: "Invalid method parameters",
+};
+
+export class ProviderError extends CustomError implements ProviderRpcError {
+  public code: number;
+  public data?: unknown;
+
+  constructor(code: ProviderErrorCode, parentError?: Error) {
```

Comment:
I think this is incorrect, as the values in `ProviderErrorCode` are just a subset of all the possible error codes. Same for the messages.

I think it would be better to receive the code and message independently.

For example, the HTTP provider initializes the provider errors with a code returned by a json-rpc server, potentially missing the message if it's the code is neither `-32005` nor `-32602`.